### PR TITLE
Add delete_protection to SQL Instances + Service Networking Pre Create Checks

### PR DESCRIPTION
**Pull request by upodroid**
Fixes: https://github.com/hashicorp/terraform-provider-google/issues/7072
Fixes: https://github.com/hashicorp/terraform-provider-google/issues/7154

If this PR is for Terraform, I acknowledge that I have:

- [X] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [X] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-downstream-tools), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [X] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/third_party/terraform/tests) (for handwritten resources or update tests).
- [X] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/master/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [X] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/master/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:note
Added `deletion_protection` to `google_sql_database_instance`, which defaults to true. SQL instances can no longer be destroyed without setting `deletion_protection = false`.
```

```release-note:enhancement
sql: added a check to `google_sql_database_instance` to catch failures early by seeing if Service Networking Connections already exists for the private network of the instance.
```

### CODE REVIEW

**Comment by upodroid** [comment]
So, tests are failing for the deletion_protection but the service networking check is passing.

```
 REDACTED  MCW0CDP3YY  ~  go  …  github.com  hashicorp  terraform-provider-google-beta   master  39⬇  5✎  USAGE  $    make testacc TEST=./google-beta TESTARGS='-run=TestAccSqlDatabaseInstance_settings_checkServiceNetworking'
==> Checking source code against gofmt...
==> Checking that code complies with gofmt requirements...
TF_ACC=1 TF_SCHEMA_PANIC_ON_ERROR=1 go test ./google-beta -v -run=TestAccSqlDatabaseInstance_settings_checkServiceNetworking -timeout 240m -ldflags="-X=github.com/hashicorp/terraform-provider-google-beta/version.ProviderVersion=acc"
=== RUN   TestAccSqlDatabaseInstance_settings_checkServiceNetworking
=== PAUSE TestAccSqlDatabaseInstance_settings_checkServiceNetworking
=== CONT  TestAccSqlDatabaseInstance_settings_checkServiceNetworking
--- PASS: TestAccSqlDatabaseInstance_settings_checkServiceNetworking (121.22s)
PASS
ok      github.com/hashicorp/terraform-provider-google-beta/google-beta 123.765s
```

The other test is complaining of non empty plan but my approach is similar to #3450 so not sure what is going on.

`make test` is also returning a long stack trace that starts with this. Seen it in #3917 as well.

```
2020/09/06 23:06:11 [TRACE] Waiting 500ms before next try
2020/09/06 23:06:12 [WARN] Composer version didn't match regexp: composer-1.4-airflow-1.9.0
fatal error: concurrent map writes

goroutine 371 [running]:
runtime.throw(0x3d33d64, 0x15)
	/usr/local/Cellar/go/1.14.4/libexec/src/runtime/panic.go:1116 +0x72 fp=0xc000b3eb38 sp=0xc000b3eb08 pc=0x1036672
runtime.mapassign_faststr(0x383c480, 0xc0002fc450, 0x3d754c3, 0x26, 0x5a30740)
	/usr/local/Cellar/go/1.14.4/libexec/src/runtime/map_faststr.go:211 +0x3f7 fp=0xc000b3eba0 sp=0xc000b3eb38 pc=0x1015997
```

File: third_party/terraform/resources/resource_sql_database_instance.go.erb
Comment (ndmckinley) [inline third_party/terraform/resources/resource_sql_database_instance.go.erb:1037]: I'm pretty sure this is best achieved with lifecycle_params: prevent_destroy - I don't think it belongs here.

File: third_party/terraform/resources/resource_sql_database_instance.go.erb
Comment (ndmckinley) [inline third_party/terraform/resources/resource_sql_database_instance.go.erb:674]: typo: "at least", and "connection".  Can you also link the docs?

**Review by ndmckinley — CHANGES_REQUESTED** [review]

File: third_party/terraform/resources/resource_sql_database_instance.go.erb
Comment (upodroid) [reply third_party/terraform/resources/resource_sql_database_instance.go.erb:1037]: There is precedent for doing this instead of using lifecycle params. https://github.com/GoogleCloudPlatform/magic-modules/pull/3450

File: third_party/terraform/resources/resource_sql_database_instance.go.erb
Comment (upodroid) [reply third_party/terraform/resources/resource_sql_database_instance.go.erb:674]: Fixed

**Comment by ndmckinley** [comment]
Were you able to figure out the `make test` stacktrace?  I'm not getting that on my end.

**Comment by ndmckinley** [comment]
Yeah, still looks good on my end.  I'm running the affected integration tests on our CI now.

**Comment by ndmckinley** [comment]
Almost there!  You didn't update the regex for your error checker in the integration test:

```
TestAccSqlDatabaseInstance_settings_checkServiceNetworking: testing.go:667: Step 0, expected error:
errors during apply: Error, failed to create instance because the network doesn't have at least 1 private services connection. Please see https://cloud.google.com/sql/docs/mysql/private-ip#network_requirements for how to create this connection.
To match:
Error, failed to create instance because the network doesn't have atleast 1 private services connections. Please review the docs on how to create this connection.
--- FAIL: TestAccSqlDatabaseInstance_settings_checkServiceNetworking (89.08s)
```

**Comment by ndmckinley** [comment]
Ah, this also breaks all the other SQL tests, because it introduces a permadiff - if deletion_protection is unset, it will always show a diff.

**Comment by upodroid** [comment]
:( I have updated the regex, any ideas on the permadiff?

**Comment by ndmckinley** [comment]
Aha.  I think it's probably an odd manifestation of the fact that you have the type set to TypeString when it should be a boolean.

**Review by ndmckinley — APPROVED** [review]
Looks good, running the tests again now.

**Comment by ndmckinley** [comment]
Okay, all the integration tests still fail with
```
Test ended in panic.
------- Stdout: -------
=== RUN   TestAccSqlDatabaseInstance_settings_basic
=== PAUSE TestAccSqlDatabaseInstance_settings_basic
=== CONT  TestAccSqlDatabaseInstance_settings_basic
------- Stderr: -------
panic: settings.0.deletion_protection: can only set full list
goroutine 937 [running]:
```

File: third_party/terraform/resources/resource_sql_database_instance.go.erb
Comment (danawillow) [inline third_party/terraform/resources/resource_sql_database_instance.go.erb:124]: why is this in settings instead of top-level on the resource?

File: third_party/terraform/resources/resource_sql_database_instance.go.erb
Comment (danawillow) [inline third_party/terraform/resources/resource_sql_database_instance.go.erb:126]: How useful is this field with a `false` default? The reason why we added it to bigtable was to prevent people from accidentally destroying their instances because they didn't realize they should do a lifecycle prevent_destroy. If the default is false, wouldn't that just be equivalent to the prevent_destroy behavior?

File: third_party/terraform/resources/resource_sql_database_instance.go.erb
Comment (danawillow) [inline third_party/terraform/resources/resource_sql_database_instance.go.erb:659]: This change is completely unrelated to the deletion protection one. In the future, changes like these two should be in separate PRs so they can be easily reviewed independently.

File: third_party/terraform/website/docs/r/sql_database_instance.html.markdown
Comment (danawillow) [inline third_party/terraform/website/docs/r/sql_database_instance.html.markdown:243]: Please add additional documentation at the top of this page, similar to https://github.com/GoogleCloudPlatform/magic-modules/pull/3450/files#diff-cabec47f1c3393e21fdded661c4b35a8. The part I think is the most important to call out explicitly is that the field has to be false **in state**, which means setting it and then running `terraform apply` so that it sticks.

Also, the field name is `deletion_protection`.

**Review by danawillow — CHANGES_REQUESTED** [review]
Apologies for the detailed drive-by- I happened to notice this PR and since I wrote the other deletion_protection one, I have some Opinions on what the right approach is :)

File: third_party/terraform/resources/resource_sql_database_instance.go.erb
Comment (upodroid) [reply third_party/terraform/resources/resource_sql_database_instance.go.erb:124]: That is where the settings are specified for sql instances. I can make it a top level key.

File: third_party/terraform/resources/resource_sql_database_instance.go.erb
Comment (upodroid) [reply third_party/terraform/resources/resource_sql_database_instance.go.erb:126]: I didn't consider that. I assumed that the users would set this true if they wanted to avoid terraform deleting their data. I never understood how the OP in 7072 blindly ran a plan that marked the instance for destruction and asked for more guardrails.

I'll set it to true.

Comment (upodroid) [reply third_party/terraform/resources/resource_sql_database_instance.go.erb:126]: In fairness, Google doesn't stop the instance from being deleted if a database already exists on it which is strange considering Google requires depedant resource to be deleted before the parent resource is deleted. ie DNS Records in DNS Zones, VPC Networks, GCE Disk, etc

**Comment by upodroid** [comment]
I fixed everything.

Ran my new test and one of the existing tests.

```
 REDACTED  MCW0CDP3YY  ~  go  …  github.com  hashicorp  terraform-provider-google-beta   master  80⬇  5✎  $    make testacc TEST=./google-beta TESTARGS='-run=TestAccSqlDatabaseInstance_settings_deletionProtection'
==> Checking source code against gofmt...
==> Checking that code complies with gofmt requirements...
TF_ACC=1 TF_SCHEMA_PANIC_ON_ERROR=1 go test ./google-beta -v -run=TestAccSqlDatabaseInstance_settings_deletionProtection -timeout 240m -ldflags="-X=github.com/hashicorp/terraform-provider-google-beta/version.ProviderVersion=acc"
=== RUN   TestAccSqlDatabaseInstance_settings_deletionProtection
=== PAUSE TestAccSqlDatabaseInstance_settings_deletionProtection
=== CONT  TestAccSqlDatabaseInstance_settings_deletionProtection
--- PASS: TestAccSqlDatabaseInstance_settings_deletionProtection (284.08s)
PASS
ok      github.com/hashicorp/terraform-provider-google-beta/google-beta 285.157s
 REDACTED  MCW0CDP3YY  ~  go  …  github.com  hashicorp  terraform-provider-google-beta   master  80⬇  5✎  USAGE  $    make testacc TEST=./google-beta TESTARGS='-run=TestAccSqlDatabaseInstance_basicSecondGen'
==> Checking source code against gofmt...
==> Checking that code complies with gofmt requirements...
TF_ACC=1 TF_SCHEMA_PANIC_ON_ERROR=1 go test ./google-beta -v -run=TestAccSqlDatabaseInstance_basicSecondGen -timeout 240m -ldflags="-X=github.com/hashicorp/terraform-provider-google-beta/version.ProviderVersion=acc"
=== RUN   TestAccSqlDatabaseInstance_basicSecondGen
=== PAUSE TestAccSqlDatabaseInstance_basicSecondGen
=== CONT  TestAccSqlDatabaseInstance_basicSecondGen
--- PASS: TestAccSqlDatabaseInstance_basicSecondGen (292.21s)
PASS
ok      github.com/hashicorp/terraform-provider-google-beta/google-beta 293.653s
```

File: third_party/terraform/website/docs/r/sql_database_instance.html.markdown
Comment (upodroid) [reply third_party/terraform/website/docs/r/sql_database_instance.html.markdown]: fixed

**Comment by ndmckinley** [comment]
Rerunning the tests now.

**Comment by ndmckinley** [comment]
The following tests still fail:

```
TestAccSqlUser_mysql
TestAccSqlClientCert_postgres
TestAccSqlUser_postgres
TestAccSqlClientCert_mysql
TestAccSqlDatabase_update
TestAccSqlDatabase_basic
```

As well as two that are already failing:
```
TestAccSqlDatabaseInstance_basicInferredName
TestAccSqlDatabaseInstance_PointInTimeRecoveryEnabled
```

**Comment by upodroid** [comment]
I forgot about the other resources linked to it. I'll push a commit now

**Comment by upodroid** [comment]
I think i got them all

**Comment by ndmckinley** [comment]
Okay, I will run the tests overnight.

**Review by ndmckinley — APPROVED** [review]
Tests all pass.  I updated the release notes as well.

**Review by ndmckinley — CHANGES_REQUESTED** [review]
Ah, one last thing:

```
==> Checking source code against linters...
google/resource_sql_database_instance.go:1062:7: Error return value of `d.Set` is not checked (errcheck)
	d.Set("deletion_protection", true)
	     ^
make: *** [lint] Error 1
```

**Comment by upodroid** [comment]
it is fixed now

**Review by ndmckinley — APPROVED** [review]

**Comment by ndmckinley** [comment]
@danawillow - are your requested changes made?  GH wants all reviewers to approve before submitting - can you approve or dismiss your review if we're good to merge?  :)

File: third_party/terraform/website/docs/r/sql_database_instance.html.markdown
Comment (danawillow) [inline third_party/terraform/website/docs/r/sql_database_instance.html.markdown:44]:
```suggestion
It is recommended to not set this field (or set it to true) until you're ready to destroy the instance and its databases.
```

**Review by danawillow — APPROVED** [review]
FYI I updated the changelog message to use the correct resource and field names.

File: third_party/terraform/website/docs/r/sql_database_instance.html.markdown
Comment (upodroid) [reply third_party/terraform/website/docs/r/sql_database_instance.html.markdown:44]: fixed

**Comment by upodroid** [comment]
I think we are ready to go.

**Comment by ndmckinley** [comment]
Not sure what changed, but `make test` now reports

```
# github.com/hashicorp/terraform-provider-google-beta/google-beta
google-beta/resource_sql_database_instance.go:666:76: not enough arguments in call to retrieveServiceNetworkingNetworkName
	have (*schema.ResourceData, *Config, string)
	want (*schema.ResourceData, *Config, string, string)
note: module requires Go 1.14
make: *** [build] Error 2
GNUmakefile:9: recipe for target 'build' failed
```

**Comment by upodroid** [comment]
something snuck in when i merged master in to the branch, i'll fix it now

**Comment by upodroid** [comment]
i think we are good now. 😌

**Pull request merged**
